### ISSUE TITLE
Support Google Cloud Spanner

### ISSUE THREAD

**Issue by jhalborg** [prisma/prisma#717]
Hey! :) 

It would be awesome if Prisma Client could be compatible with https://cloud.google.com/spanner/

**Comment by wgottschalk** [comment]
I'd love to see Spanner supported. Now that Spanner has [granular instance sizing](https://cloud.google.com/blog/products/databases/get-more-out-of-spanner-with-granular-instance-sizing)[](https://github.com/wgottschalk) and cloud sql has mandatory downtime where Spanner doesn't, it makes it the preferred database over cloudsql.
 
Spanner is starting to rollout a postgres dialect to spanner. They mention it's not 100% compatible but it would be great have prisma add some documentation outlining limitations if any

https://cloud.google.com/spanner/docs/postgresql-interface

**Comment by janpio** [comment]
Nice.

(Note that the following does not indicate that we are committed to build a Google Spanner compatible connector any time soon. This is just me exploring and figuring out what might already work, and what does not. If Prisma will ever have this, and wants to spend more time on this, will be decided on a different level. I am just curious.)

Our next steps here would be to understand GCP enough to be able to have a Spanner instance with some schema and data in it, and enable the PostgreSQL interface. We would then try to run `npx prisma db pull` on it and see what fails (If it does not, we would probably still continue with a manually written schema of the database - starting with one table probably.) If it magically works out of the box, we would use the generated schema to send some manualy queries. If that also magically works, we could even try `migrate dev` after a small schema change to see what happens. And if _that_ also succeeds we would then start to think how to integrate a Cloud only service into our test suite - but that is quite a bit off.

So if _you_ understand GCP well, have a Spanner instance with data and the PostgreSQL interface enabled - please try these steps! (And/or send me an email at jan@prisma.io and send over the connection string, so we can do so at the side).

**Comment by glebbash** [comment]
Tried to use [PGAdapter](https://cloud.google.com/spanner/docs/pgadapter) and [Spanner Emulator](https://cloud.google.com/spanner/docs/emulator) locally with Prisma using the following docker-compose file:

```yaml
version: "3.3"
services:
  spanner:
    env_file: .env
    image: gcr.io/cloud-spanner-emulator/emulator
    ports:
      - '9010:9010'
      - '9020:9020'

  pg-adapter:
    image: gcr.io/cloud-spanner-pg-adapter/pgadapter:v0.6.1
    volumes:
      - ./test/fake-spanner-credentials.json:/acct_credentials.json
    ports:
      - 127.0.0.1:5432:5432
    command:
      - -p my_project
      - -i my_instance
      - -d my_database
      - -c /acct_credentials.json
      - -x
```

Getting this error:

```
Prisma CLI Version : 4.1.1
Error: Command failed: DATABASE_URL=cloudspanner:/projects/my_project/instances/my_instance/databases/my_database prisma migrate dev
Error: Get Config: Schema parsing - Error while interacting with query-engine-node-api library
Error code: P1012
error: Error validating datasource `db`: the URL must start with the protocol `postgresql://` or `postgres://`.
  -->  schema.prisma:6
   | 
 5 |   provider = "postgresql"
 6 |   url      = env("DATABASE_URL")
```

So the problem is that PGAdapter requires connection url to start with `cloudspanner://` while Prisma wants it to be `postgresql://`. 

Using `cloudspanner://` will error out on Prisma side.

Using `postgresql://` will error out on PGAdapter side.

**Comment by janpio** [comment]
Adding an env var that allows `cloudspanner://` as an alternative to `postgresql://` (in a `dev` version maybe) would unblock this problem I guess @glebbash?

**Comment by glebbash** [comment]
If there is no further validation for the rest of connection url: `/projects/my_project/instances/my_instance/databases/my_database` then yeah, it should work.

**Comment by janpio** [comment]
Thanks for clarifying that, we would indeed probably check for port and so on. So they way here would be to accept a completely different connection string. Interesting that `psql` seems to work for that out of the box (per PGAdapter docs). Can you share a full connection string that we would need to use?

**Comment by glebbash** [comment]
Now that I think about which url would be okay for PGAdapter to work I am kinda lost, as it requires cloudspanner prefix and after that you would need to point to a local IP and port.

So theoretically it should be comething like `cloudspanner://localhost:5432`.

But there is a big chance that it wont work like that as PGAdapter docs say that only `psql` client is supported.

**Comment by NOT-HAL9000** [comment]
Has there been any updates / further thoughts on this @janpio? 

Prisma should seriously consider supporting a Spanner connector.

The [PostgreSQL interface for Cloud Spanner is now in GA](https://www.infoq.com/news/2022/07/google-cloud-spanner-postgresql/) and their basic pricing tier has dropped down to as [low as $40/month](https://cloud.google.com/blog/products/databases/use-spanner-at-low-cost-with-granular-instance-sizing). 

Having a cheap, [strongly consistent](https://cloud.google.com/spanner/docs/true-time-external-consistency) cross-region, ACID compliant, relational database is basically the holy grail, but unfortunately, many fast-moving teams will either overlook or avoid using Spanner because there is no Type-safe Node ORM that connects to it (despite the efforts by Google to provide the PostgreSQL Interface).

By supporting Spanner, Prisma has the opportunity to capture the market of large enterprise companies who use and/or want to use it with Node. This could then be leveraged as a funnel to drive these large customers towards your paid and enterprise offerings, which pays off the effort spent integrating Spanner.

**Comment by janpio** [comment]
This is currently not a priority for us. 

If someone (@glebbash?) confirms that for example allowing `cloudspanner://` as a protocol is all it needs to test this further, we might be happy to add that so we get to a better understanding where blockers will be. But w can currently not invest in figuring all this out or even implementing support unfortunately.

**Comment by etifontaine** [comment]
Hello,

I have setup a Spanner instance, with the `PostgreSQL` dialect.

When starting my app, I get the following error:
`Error querying the database: db error: ERROR: Invalid SET statement: SET NAMES 'UTF8'. Expected TO or =.`

I use `DATABASE_URL="postgresql://127.0.0.1:5432/my-database"` in my Prisma config.

Plus the following Docker container,
`docker run \
  -d -p 5432:5432 \
  -v ${GOOGLE_APPLICATION_CREDENTIALS}:${GOOGLE_APPLICATION_CREDENTIALS}:ro \
  -e GOOGLE_APPLICATION_CREDENTIALS \
  gcr.io/cloud-spanner-pg-adapter/pgadapter \
  -p my-gcp-project-id -i my-spanner-instance-id \
  -x`

**Comment by janpio** [comment]
Naive question, what is the quickest way to setup up the required `GOOGLE_APPLICATION_CREDENTIALS`?

**Comment by NOT-HAL9000** [comment]
I just tried the same thing as @etifontaine and got the same result.

1. Create a Google Cloud Account
2. Set up a Regional Spanner instance with 100 processing units ($0.09/h)
3. When it's done, create a database with PG Dialect. Remember the name.
4. Go to IAM > Services Accounts and create a new one with Owner Access. 
5. Once created, click the 3 dots next to the new service account and 'Manage Keys'
6. Create a new JSON key, this will be your GOOGLE_APPLICATION_CREDENTIALS.
7. Follow the guide here and launch a local PGAdapter docker service. Below is the what I used. [PROJECT_ID](https://updraftplus.com/faqs/where-do-i-find-my-google-project-id/) is found here. Spanner Instance ID is found under Spanner Tab and the database name is what you named it in step 3. CREDENTIALS_FILE_PATH is the downloaded JSON path from step 5. 

```
docker run -d -p 5432:5432 \
    -v CREDENTIALS_FILE_PATH:/acct_credentials.json \
    gcr.io/cloud-spanner-pg-adapter/pgadapter:latest \
    -p PROJECT_ID -i SPANNER_INSTANCE_ID -d PG_DATABASE_NAME  \
    -c /acct_credentials.json -q -x
```

8. Create a basic node project with with the following
```
// prisma/schema.prisma
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}
```
```
// .env
DATABASE_URL="postgresql://localhost:5432/{DATABASE_NAME from step 3}"
```

```
// index.js
const { PrismaClient } = require('@prisma/client')
const prisma = new PrismaClient();

(async () => {
    const result = await prisma.$queryRaw`SELECT 1`
    console.log(result)
})()
```

`Error querying the database: db error: ERROR: Invalid SET statement: SET NAMES 'UTF8'. Expected TO or =.`

I also got the same error when trying. 

```npx prisma migrate dev --name init```

-- 
When finished testing, make sure to delete your spanner instance. There is also a free trial for Spanner, but I have already used it. So testing for you may be free.

**Comment by janpio** [comment]
Some experimental work started over there: https://github.com/GoogleCloudPlatform/pgadapter/issues/589 + https://github.com/GoogleCloudPlatform/pgadapter/pull/590 <3

**Comment by patwhite** [comment]
I'll add my two cents here - it would actually be nice to have a spanner native (non-postgres dialect) version. There are a few really interesting caveats to the way the postgres dialect works as have been noted in this thread. First, you have to run a separate process (the pgadapter) to do the translation, but secondly, there are actually some really important features missing when you use the postgres dialect - specifically, the change log that spanner produces and is consumable for CDC doesn't work with postgres (as I understand it). I also remember the journey we went through here for cockroach where, even though it said wire compatible, there are enough special things that prisma does that it didn't really work perfectly until there was a full, flavored release.

**Comment by janpio** [comment]
We fundamentally agree, but to explain why I am pushing on this: Native Spanner support is _not_ happening for sure this year. If it turns out that the PgAdapter is close enough or can be adapted though, we could possibly find a way to make this usable sooner.

**Comment by janpio** [comment]
I think `version()` again is something the adapter could handle. It is one of the default system functions of PostgreSQL, which is why Prisma expects it to exist: https://www.postgresql.org/docs/current/functions-info.html#:~:text=to%20current_user.-,version%20()%20%E2%86%92%20text,-Returns%20a%20string 

(We _could_ rewrite our code to handle this not being present, but for normal PostgreSQL (and compatible databases) that is never expected to fail, so would be a bit weird - but we most probably _could_: https://github.com/prisma/prisma-engines/blob/af46365b40aa4864b99acec97dc2dd5bdc27411c/migration-engine/connectors/sql-migration-connector/src/flavour/postgres.rs#L562-L594)

I expect a few more of these to pop up on `db pull` as well, where Prisma queries a lot of the PostgreSQL system _tables_ for information about the available tables and columns and so on.

**Comment by NOT-HAL9000** [comment]
I've mentioned this on the prisma support issue: https://github.com/GoogleCloudPlatform/pgadapter/issues/589

**Comment by jillxuu** [comment]
im running into similar errors when using pgadapter with prisma migrate
```
Error: Schema engine error:
db error: ERROR: Postgres function current_setting(text) is not supported - Statement: 'with pg_namespace as (
  select case schema_name when 'pg_catalog' then 11 when 'public' then 2200 else 0 end as oid,
        schema_name as nspname, null as nspowner, null as nspacl
  from information_schema.schemata
)
SELECT EXISTS(SELECT 1 FROM pg_namespace WHERE nspname = $1), '14.1', current_setting('server_version_num')::integer as numeric_version;'
```

**Comment by jillxuu** [comment]
fyi, i changed my url to be `DATABASE_URL="postgres://postgres@127.0.0.1:5432/test_db?schema=transactions"` which includes both the database name as well as the schema name, and `npx prisma migrate dev --name init` is able to run successfully.

**Comment by fuvidani** [comment]
Hello everyone, do you have any update on the Spanner support? Has anybody been able to make it work reliably with Prisma? 🙇‍♂️

**Comment by m3adow** [comment]
We recently tried to use Spanner as a Prisma backend but gave it up in the end. If I remember correctly, we couldn't make migrations work properly without more work than it was worth.

I'd love to hear if and how anyone got this to work properly

**Comment by igalklebanov** [comment]
Hey 👋

Closing this for being dusty and stale.
